content/blog/2019-11-13-null-pointer-guards.md

```diff
@@ -0,0 +1,220 @@
+which has various benchmarks for different languages. I chose a small portion of
+these tests to run with and without my compiler passes.
+
+I chose to run:
+- [binary-trees](https://benchmarksgame-team.pages.debian.net/benchmarksgame/program/binarytrees-gpp-2.html)
+  - Ran with: `./binary-trees 13`
+- [fannkuch-redux](https://benchmarksgame-team.pages.debian.net/benchmarksgame/program/fannkuchredux-gpp-5.html)
+  - Ran with: `./fannkucuh-redux 10`
+- [fasta](https://benchmarksgame-team.pages.debian.net/benchmarksgame/program/fasta-gpp-1.html)
+  - Ran with: `./fasta 250000`
+- [mandelbrot](https://benchmarksgame-team.pages.debian.net/benchmarksgame/program/mandelbrot-gpp-5.html)
+  - Ran with: `./mandelbrot 2000`
+- [n-body](https://benchmarksgame-team.pages.debian.net/benchmarksgame/program/nbody-gpp-1.html)
+  - Ran with: `./nbody 5000000`
+
+Here we simply show the mean time taken to run these programs and the standard
+deviation. There are three columns, representing the three different ways we
+compiled the program.
+- *Baseline* is just compiling using Clang without our null check passes.
+Compiled with `/usr/local/opt/llvm/bin/clang++ -O2 <benchmark>.cpp -o <benchmark>`, e.g.,
+`/usr/local/opt/llvm/bin/clang++ -O2 binary-trees.cpp -o binary-trees`
+- *No NPA* is compiling using the null check passes, but without taking into
+consideration which pointers are `PossiblyNull` or `DefinitelyNonNull`. That is,
+we do a null check on every pointer dereference.
+NOTE: Had to go into source code and remove the check for if pointers are `PossiblyNull`
+Compiled with `../tests/compile-with-opt.sh nbody.cpp -O2`
+- *With NPA* is compiling using the null check passes *and* eliding null checks
+if the pointer is `DefinitelyNonNull`.
+Compiled with same instruction as above.
+
+**Runtime Means and Standard Deviations**
+
+|                 |      Baseline      |      No NPA      |     With NPA     |
+|:---------------:|:------------------:|:----------------:|:----------------:|
+|  binary-trees   |  208 ms ± 4.78 ms  | 217 ms ± 12.8 ms | 213 ms ± 6.78 ms |
+|  fannkuch-redux |  222 ms ± 5.02 ms  | 536 ms ± 6.34 ms | 563 ms ± 54.9 ms |
+|  fasta          |  200 ms ± 6.01 ms  | 365 ms ± 7.77 ms | 375 ms ± 28.8 ms |
+|  mandelbrot     |  319 ms ± 5.79 ms  | 294 ms ± 2.52 ms | 298 ms ± 9.76 ms |
+|  n-body         |  327 ms ± 16.7 ms  | 733 ms ± 12.4 ms | 734 ms ± 17.8 ms |
+
+Here we can see that in majority of cases, the baseline performs significantly
+better than the programs with null checks. To me this is a little surprising
+because I figured the branch predictor would always know that the null checks
+don't do anything except return from the null checking function.
+
+It is also interesting to note that the NPA didn't cause any improvement! This
+may be because not that many null pointer checks were removed, perhaps because
+the analysis is too conservative; this would require further investigation.
```

Comment:
Any chance you have a rough count of the number of (static) checks that were removed?

Reply:
Nope :/